File: NetFramework/XHtmlEditor/VB/C1EditorQuickStart/ToolStrips/ToolStripBase.vb
Imports System.Drawing
Imports System.Windows.Forms
Imports System.Collections.Generic
Imports System.Text
Imports C1.Win.C1Editor

Namespace ToolStrips
    MustInherit Class ToolStripBase
        Inherits System.Windows.Forms.ToolStrip
        ' editor connected to this toolstrip
        Private _editor As C1Editor

        ' ctor
        Public Sub New()
            MyBase.New()
            OnInitialize()
        End Sub

        ' notify of changes to the editor content
        Public Property Editor() As C1.Win.C1Editor.C1Editor
            Get
                Return _editor
            End Get
            Set(ByVal value As C1.Win.C1Editor.C1Editor)
                If Editor IsNot Nothing Then
                    RemoveHandler Editor.SelectionChanged, AddressOf Editor_SelectionChanged
                    RemoveHandler Editor.DocumentChanged, AddressOf Editor_DocumentChanged
                    RemoveHandler Editor.ModeChanged, AddressOf Editor_ModeChanged
                    RemoveHandler Editor.KeyDown, AddressOf Editor_KeyDown
                End If

                _editor = value

                If Editor IsNot Nothing Then
                    AddHandler Editor.SelectionChanged, AddressOf Editor_SelectionChanged
                    AddHandler Editor.DocumentChanged, AddressOf Editor_DocumentChanged
                    AddHandler Editor.ModeChanged, AddressOf Editor_ModeChanged
                    AddHandler Editor.KeyDown, AddressOf Editor_KeyDown
                End If
            End Set
        End Property
        Private Sub Editor_DocumentChanged(ByVal sender As Object, ByVal e As EventArgs)
            OnDocumentChanged(e)
        End Sub
        Private Sub Editor_SelectionChanged(ByVal sender As Object, ByVal e As EventArgs)
            OnSelectionChanged(e)
        End Sub
        Private Sub Editor_ModeChanged(ByVal sender As Object, ByVal e As EventArgs)
            OnModeChanged(e)
        End Sub
        Private Sub Editor_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs)
            OnEditorKeyDownHandler(e)
        End Sub
        Protected Overridable Sub OnDocumentChanged(ByVal e As EventArgs)
        End Sub
        Protected Overridable Sub OnSelectionChanged(ByVal e As EventArgs)
        End Sub
        Protected Overridable Sub OnModeChanged(ByVal e As EventArgs)
        End Sub
        Protected Overridable Sub OnEditorKeyDownHandler(ByVal e As KeyEventArgs)
        End Sub

        ' initialize toolstrip
        Protected Overridable Sub OnInitialize()
        End Sub

        ' enable/disable toolbar buttons
        Public Overridable Sub UpdateState()
        End Sub

        ' helper for populating the toolstrip
        Protected Function AddButton(ByVal text As String, ByVal image As Image, ByVal handler As EventHandler) As ToolStripButton
            Dim btn As ToolStripButton = New ToolStripButton(text, image)
            btn.DisplayStyle = ToolStripItemDisplayStyle.Image
            AddHandler btn.Click, handler
            Items.Add(btn)
            Return btn
        End Function
    End Class
End Namespace

File: NetFramework/XHtmlEditor/CS/C1EditorQuickStart/ToolStrips/ToolStripBase.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using System.Collections.Generic;
using System.Text;
using C1.Win.C1Editor;

namespace C1EditorQuickStart.ToolStrips
{
    abstract class ToolStripBase : System.Windows.Forms.ToolStrip
    {
        // editor connected to this toolstrip
        C1Editor _editor;

        // ctor
        public ToolStripBase()
        {
            OnInitialize();
        }

        // notify of changes to the editor content
        public C1.Win.C1Editor.C1Editor Editor
        {
            get { return _editor; }
            set 
            {
                if (Editor != null)
                {
                    Editor.SelectionChanged -= Editor_SelectionChanged;
                    Editor.DocumentChanged -= Editor_DocumentChanged;
                    Editor.ModeChanged -= Editor_ModeChanged;
                    Editor.KeyDown -= Editor_KeyDown;
                }

                _editor = value;

                if (Editor != null)
                {
                    Editor.SelectionChanged += Editor_SelectionChanged;
                    Editor.DocumentChanged += Editor_DocumentChanged;
                    Editor.ModeChanged += Editor_ModeChanged;
                    Editor.KeyDown += Editor_KeyDown;
                }
            }
        }
        void Editor_DocumentChanged(object sender, EventArgs e)
        {
            OnDocumentChanged(e);
        }
        
        void Editor_SelectionChanged(object sender, EventArgs e)
        {
            OnSelectionChanged(e);
        }

        void Editor_ModeChanged(object sender, EventArgs e)
        {
            OnModeChanged(e);
        }

        void Editor_KeyDown(object sender, KeyEventArgs e)
        {
            OnEditorKeyDownHandler(e);
        }

        protected virtual void OnDocumentChanged(EventArgs e)
        {
        }
        
        protected virtual void OnSelectionChanged(EventArgs e)
        {
        }

        protected virtual void OnModeChanged(EventArgs e)
        {
        }

        protected virtual void OnEditorKeyDownHandler(KeyEventArgs e)
        {
        }

        // initialize toolstrip
        protected virtual void OnInitialize()
        {
        }

        // enable/disable toolbar buttons
        public virtual void UpdateState()
        {
        }

        // helper for populating the toolstrip
        protected ToolStripButton AddButton(string text, Image image, EventHandler handler)
        {
            ToolStripButton btn = new ToolStripButton(text, image);
            btn.DisplayStyle = ToolStripItemDisplayStyle.Image;
            btn.Click += handler;
            Items.Add(btn);
            return btn;
        }
    }
}
